README.md
# HawkScan

![alt tag](https://github.com/c0dejump/HawkScan/blob/master/static/logo.jpg)

Security Tool for Reconnaissance and Information Gathering on a website. (python 2.x & 3.x)

This script use "WafW00f" to detect the WAF in the first step (https://github.com/EnableSecurity/wafw00f)

This script use "Sublist3r" to scan subdomains (https://github.com/aboul3la/Sublist3r)

This script use "waybacktool" to check in waybackmachine (https://github.com/Rhynorater/waybacktool)

# News
**!** V 1.3.3 !   
**!** Replace "--cookie" by "-H" for different header values; ex: -H "Host:test" // -H "Authentification:cookie" (not space after ":" or "=")        
**!** Adding new function which try automatically if it's possible scanning with "localhost" host     

# Features
 - [x] URL fuzzing and dir/file detection
 - [x] Test backup/old file on all the files found (index.php.bak, index.php~ ...)
 - [x] Check header information
 - [x] Check DNS information
 - [x] Check whois information
 - [x] User-agent random or personal
 - [x] Extract files
 - [x] Keep a trace of the scan
 - [x] Check @mail in the website and check if @mails leaked
 - [x] CMS detection + version and vulns
 - [x] Subdomain Checker
 - [x] Backup system (if the script stopped, it take again in same place)
 - [x] WAF detection
 - [x] Add personal prefix
 - [x] Auto update script
 - [x] Auto or personal output of scan (scan.txt)
 - [x] Check Github
 - [x] Recursif dir/file
 - [x] Scan with an authenfication cookie
 - [x] Option --profil to pass profil page during the scan
 - [x] HTML report
 - [x] Work it with py2 and py3
 - [x] Add option rate-limit if app is unstable (--timesleep)
 - [x] Check in waybackmachine
 - [x] Response error to WAF
 - [x] Check if DataBase firebaseio existe and accessible
 - [x] Automatic threads depending response to website (and reconfig if WAF detected too many times). Max: 30
 - [x] Search S3 buckets in source code page
 - [x] Testing bypass of waf if detected
 - [x] Testing if it's possible scanning with "localhost" host
 
# TODO 
**P1 is the most important**

 - [ ] Dockerfile [P1]
 - [ ] JS parsing and analysis [P1]
 - [ ] Analyse html code webpage [P1]
 - [ ] On-the-fly writing report [P1]
 - [ ] Check HTTP headers/ssl security [P2]
 - [ ] Fuzzing amazonaws S3 Buckets [P2]
 - [ ] Anonymous routing through some proxy (http/s proxy list) [P2]
 - [ ] Check pastebin [P2]
 - [ ] Access token [P2]
 - [ ] Check source code and verify leak or sentsitive data in the Github [P2]
 - [ ] Check phpmyadmin version [P3]
 - [ ] Scan API endpoints/informations leaks [ASAP]
 
 # Usage
 > 
 
       pip(3) install -r requirements.txt 
    If problem with pip3:    
       sudo python3 -m pip install -r requirements.txt
 > 
  
 >
     
    usage: hawkscan.py [-h] [-u URL] [-w WORDLIST] [-s SUBDOMAINS] [-t THREAD] [-a USER_AGENT] [--redirect] [-r] [-p PREFIX] [-o OUTPUT] [--cookie COOKIE_] [--exclude EXCLUDE] [--timesleep TS] [--auto]
 
 > 
 
    optional arguments: 
     -h, --help         show this help message and exit
     -u URL             URL to scan [required]
     -w WORDLIST        Wordlist used for URL Fuzzing. Default: dico.txt
     -s SUBDOMAINS      Subdomain tester
     -t THREAD          Number of threads to use for URL Fuzzing. Default: 20
     -a USER_AGENT      Choice user-agent 
     --redirect         For scan with redirect response (301/302) 
     -r                 Recursive dir/files      
     -p PREFIX          Add prefix in wordlist to scan      
     -o OUTPUT          Output to site_scan.txt (default in website directory)       
     -b                 Add a backup file scan like 'exemple.com/~exemple/, exemple.com/ex.php.bak...' but longer             
     -H HEADER_         modify HEADER              
     --exclude EXCLUDE  To define a page or response code status type to exclude during scan                                            
     --timesleep TS     To define a timesleep/rate-limit if app is unstable during scan                                 
     --auto             Automatic threads depending response to website. Max: 30      
     --update           For automatic update

 >

# Exemples

 >
    //Basic
    python hawkscan.py -u https://www.exemple.com -w dico_extra.txt

    //With redirect
    python hawkscan.py -u https://www.exemple.com -w dico_extra.txt -t 5 --redirect

    //With backup files scan
    python hawkscan.py -u https://www.exemple.com -w dico_extra.txt -t 5 -b

    //With an exclude page
    python hawkscan.py -u https://www.exemple.com -w dico_extra.txt -t 5 --exclude https://www.exemple.com/profile.php?id=1

    //With an exclude response code
     python hawkscan.py -u https://www.exemple.com -w dico_extra.txt -t 5 --exclude 403

 >

# Thanks
Layno (https://github.com/Clayno/)      
Sanguinarius (https://twitter.com/sanguinarius_Bt)        
Cyber_Ph4ntoM (https://twitter.com/__PH4NTOM__)  


CHANGELOG.md
Changelog:
----------

	- 1.3.3
		Adding new function which try automatically if it's possible scanning with "localhost" host

	- 1.3.2
		Replace "--cookie" by "-H" for different header values; ex: -H "Host:test" // -H "Authentification:cookie" (not space after ":" or "=")

	- 1.3.1:
		Code review
		New logo
		Adding Changelog

	- 1.2:
		Adding news words in dico.txt (old dico_extra.txt)
		Adding extensions in backup check test function, option -b (.json, .xml, .bkp...) => very long
		Test bypass of waf rate limited in real time (X-Originating-IP...)
		Exclude response http code (--exclude 403)
		Filter on response http code in report

	- 1.0:
	  	Better management Threads
		Add news words in dico_extra.txt
		New style for the report
		Errors log management


modules/check_cms.py
import json
import requests
from bs4 import BeautifulSoup
from config import PLUS, WARNING, INFO, LESS, LINE

class check_cms:

    def detect_cms(self, url, directory):
        """
        CMS:
        Detect if the website use a CMS
        """
        print(INFO + "CMS")
        print(LINE)
        req = requests.get("https://whatcms.org/APIEndpoint/Detect?key=1481ff2f874c4942a734d9c499c22b6d8533007dd1f7005c586ea04efab2a3277cc8f2&url={}".format(url))
        if "Not Found" in req.text:
            with open(directory + "/cms.txt", "w+") as cms_write:
                cms_write.write("  this website does not seem to use a CMS")
            print("\t{} this website does not seem to use a CMS \n".format(LESS))
            print(LINE)
            return False, False;
        else:
            reqt = json.loads(req.text)
            result = reqt["result"].get("name")
            v = reqt["result"].get("version")
            if v:
                with open(directory + "/cms.txt", "w+") as cms_write:
                    cms_write.write("This website use {} {}".format(result, v))
                print("\t{} This website use \033[32m{} {} \033[0m\n".format(PLUS, result, v))
                return result, v;
            else:
                with open(directory + "/cms.txt", "w+") as cms_write:
                    cms_write.write("  This website use {} but nothing version found".format(result))
                print("\t{} This website use \033[32m{}\033[0m but nothing version found \n".format(PLUS, result))
                print(LINE)
                return False, False;

    def cve_cms(self, result, v):
        """
        CVE_CMS:
        Check CVE with cms and version detected by the function 'detect_cms'.
        """
        url_comp = "https://www.cvedetails.com/version-search.php?vendor={}&product=&version={}".format(result, v)
        req = requests.get(url_comp, allow_redirects=True, verify=False)
        if not "matches" in req.text:
            print("{}CVE found ! \n{}{}\n".format(WARNING, WARNING, url_comp))
            if 'WordPress' in req.text:
                version =  v.replace('.','')
                site = "https://wpvulndb.com/wordpresses/{}".format(version)
                req = requests.get(site)
                soup = BeautifulSoup(req.text, "html.parser")
                search = soup.find_all('tr')
                if search:
                    for p in search:
                        dates = p.find("td").text.strip()
                        detail = p.find("a").text.strip()
                        print("  {}{} : {}".format(WARNING, dates, detail))
                else:
                    print("  {} Nothing wpvunldb found \n".format(LESS))
        elif 'WordPress' in req.text:
            version =  v.replace('.','')
            site = "https://wpvulndb.com/wordpresses/{}".format(version)
            req = requests.get(site)
            soup = BeautifulSoup(req.text, "html.parser")
            search = soup.find_all('tr')
            if search:
                print("{}CVE found ! \n{}{}\n".format(WARNING, WARNING, site))
                for p in search:
                    dates = p.find("td").text.strip()
                    detail = p.find("a").text.strip()
                    print("{}{} : {}".format(WARNING, dates, detail))
                    print(LINE)
            else:
                print("  {} Nothing wpvunldb found \n".format(LESS))
                print(LINE)
        else:
            print("  {} Nothing CVE found \n".format(LESS))
            print(LINE)

modules/bypass_waf.py
import requests
import socket
import traceback
from modules.detect_waf import verify_waf
from config import PLUS, WARNING, INFO

requests.packages.urllib3.disable_warnings(requests.packages.urllib3.exceptions.InsecureRequestWarning)

def bypass_waf(req, res):
	"""
	Bypass_waf: try if the waf can be bypass, using different payloads
	"""
	win = False
	domain = res.split("/")[2]
	website_ip = socket.gethostbyname(domain) # take ip website
	header_base = [
	"X-Originating-IP", 
	"X-Forwarded-For",
	"X-Forwarded-Host",
	"X-Remote-IP",
	"X-Remote-Addr",
	"X-Client-IP",
	"Cookie",
	"Access-Control-Allow-Origin",
	"Origin",
	"Timing-Allow-Origin"
	]
	options = [website_ip, domain, "127.0.0.1", "*"]
	for hb in header_base:
		for o in options:
			headers = {
				hb : o
			}
			try:
				display = False
				vrfy = verify_waf(req, res, headers, display)
				#print(vrfy)
				if vrfy == False:
					#win = True
					print("{} You can bypass the WAF with: {} !".format(PLUS, headers))
					return headers
			except Exception:
				traceback.print_exc()
	if not win:
		try:
			headers = {
			"Clear-Site-Data":"*"
			}
			display = False
			vrfy = verify_waf(req, res, headers, display)
			if vrfy == False:
				#win = True
				print("{} You can bypass the WAF with: {} !".format(PLUS, headers))
				return headers
		except:
			traceback.print_exc()
		#print("\033[31m[:(]\033[0m Our tests not bypass it, sorry")
	return win

"""if __name__ == '__main__':
	req = "plop"
	user_agent = False
	res = "https://transferwise.com/"
	bypass_waf(req, res)""" #DEBUG

report/creat_report.py
from datetime import datetime
import csv
import traceback

def create_report(directory, cookie_):
    """
    Create_report: make a html report with url, waf, email...
    """
    urls = ""
    waf = ""
    mails = ""
    nowdate = datetime.now()
    nowdate = "{}-{}-{}".format(nowdate.day, nowdate.month, nowdate.year)
    #directory = "../" + directory#DEBUG
    if cookie_:
        auth_stat = "Authenticated"
    else:
        auth_stat = "No Authenticated"
    with open("report/report_"+ directory.split("/")[-1] + ".html", "a+") as test:
        with open(directory + "/scan.txt", "r") as scan:
            for s in scan.read().splitlines():
                s = s.split(" ")
                s0 = s[0]
                s1 = s[1]
                if s0 == "[+]":
                    if "301" in s or "302" in s:
                        if s[2] == "301":
                            s0 = s0.replace("[+]", "301")
                        elif s[2] == "302":
                            s0 = s0.replace("[+]", "302")
                        urls += """
                            <tr style="display:none;" class="value300">
                            <td style="color: orange; ">{}</td>
                            <td style="color: orange; "><a href="{}" target="_blank" style="color: white;">{}</a></td>
                            <td style="color: orange; ">{}</td>
                            </tr>
                            """.format(nowdate, s1, s1, s0)
                    else:
                        s0 = s0.replace("[+]", "200")
                        urls += """
                        <tr style="display:none;" class="value200">
                        <td style="color: green;">{}</td>
                        <td style="color: green;"><a href="{}" target="_blank" style="color: white;">{}</td>
                        <td style="color: green;">{}</td>
                        </tr>
                        """.format(nowdate, s1, s1, s0)
                elif s0 == "[x]":
                    s0 = s0.replace("[x]", "403")
                    urls += """
                        <tr style="display:none;" class="value403">
                        <td style="color: red;">{}</td>
                        <td style="color: red;"><a href="{}" target="_blank" style="color: white;">{}</a></td>
                        <td style="color: red;">{}</td>
                        </tr>
                        """.format(nowdate, s1, s1, s0)
                elif s0 == "[-]":
                    if "401" in s:
                        if s[2] == "401":
                            s0 = s0.replace("[-]","401")
                        urls += """
                            <tr style="display:none;" class="value401">
                            <td style="color: orange; ">{}</td>
                            <td style="color: orange; "><a href="{}" target="_blank" style="color: white;">{}</a></td>
                            <td style="color: orange; ">{}</td>
                            </tr>
                            """.format(nowdate, s1, s1, s0)
                elif s0 == "[!]":
                    if "400" in s:
                        if s[2] == "400":
                            s0 = s0.replace("[!]","400")
                        urls += """
                            <tr style="display:none;" class="value400">
                            <td style="color: red; ">{}</td>
                            <td style="color: red; "><a href="{}" target="_blank" style="color: white;">{}</a></td>
                            <td style="color: red; ">{}</td>
                            </tr>
                            """.format(nowdate, s1, s1, s0)
        try:
            with open(directory + "/waf.txt", "r") as waff:
                waf_res = ""
                for w in waff.read().splitlines():
                    if "The site" in w:
                        waf_res = w
                if waf_res:
                    waf += """
                        <span class="subText" style='color: red;'>{}</span>
                    """.format(waf_res)
        except:
            waf += """
                <span class="subText" style='color: green;'>This site dosn't seem to use a WAF</span>
            """
        try:
            with open(directory + "/mail.csv", "r") as csvFile:
                reader = csv.reader(csvFile)
                for row in reader:
                    mail = row[0]
                    stat = row[1]
                    if "no" in stat:
                        mails += """
                            <tr>
                            <td style="color: green; ">{}</td>
                            <td style="color: green; ">{}</td>
                            </tr>
                            """.format(mail, stat)
                    else:
                        mails += """
                            <tr>
                            <td style="color: red; ">{}</td>
                            <td style="color: red; ">{}</td>
                            </tr>
                            """.format(mail, stat)
        except:
            mails = "<tr><td><b> No emails found </b></td></tr>"
        try:
            link = ""
            with open(directory + "/links.txt", "r") as links:
                for l in links.read().splitlines():
                    link += """
                        <tr>
                        <td><a href="{}" target="_blank" style="color: white;">{}</a></td>
                        </tr>
                        """.format(l, l)
        except:
            link = "<tr><td> No links found </td></tr>"
        try:
            wayback = ""
            with open(directory + "/wayback.txt", "r") as waybacks:
                for wb in waybacks.read().splitlines():
                    w = wb.split(",")
                    w_status = w[1]
                    wayback += """
                        <tr>
                        <td><a href="{}">{}</a></td>
                        <td>{}</td>
                        </tr>
                        """.format(w[0], w[0], w_status)
        except:
            pass
        try:
            with open(directory + "/cms.txt","r") as cmsFile:
                cms = ""
                for cms_read in cmsFile.read().splitlines():
                    cms += """
                        <span class='subText' color: green; ">{}</span>
                        """.format(cms_read)
        except:
            cms = "<span class='subText' style='color: red;'> This site dosn't seem to use a CMS </span>"
        test.write('''
            <!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8" />
        <title>Hawkscan Report</title>
        <link href="charte.css" rel="stylesheet">
        <link href="page.css" rel="stylesheet">
        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script>
        <script src="scripts/scripts.js"></script>
    </head>

    <body>
        
        <header class="title w100">
            <div class="container">
                <h1>Hawkscan Report</h1>
            </div>
        </header>

        <main>
            <div class="subTitle container">
                <div class="w100 flex flex-jsb">
                    <div class="w1-3 centerText flex flex-dc flex-aic flex-jcc">
                        <h2>WAF</h2>
                        {}
                    </div>
                    <div class="w1-3 centerText flex flex-dc flex-aic flex-jcc">
                        <h2>CMS</h2>
                        {}
                    </div>
                    <div class="w1-3 centerText flex flex-dc flex-aic flex-jcc">
                        <h2>Status</h2>
                        <span class="subText">{}</span>
                    </div>
                </div>
            </div>
            <div class="subLinkSection w100">
                <div class="subLinkBloc w100">
                    <input type="radio" id="s1" name="s" checked/>
                    <input type="radio" id="s2" name="s"/>
                    <input type="radio" id="s3" name="s"/>
                    <input type="radio" id="s4" name="s"/>
                    <div class="subLink">
                        <div class="tabs container flex flex-jsb">
                            <label class="w1-4 centerText" for="s1">URLs</label>
                            <label class="w1-4 centerText" for="s2">Mails</label>
                            <label class="w1-4 centerText" for="s3">Links</label>
                            <label class="w1-4 centerText" for="s4">Wayback</label>
                        </div>
                    </div>
                    <ul class="sections container">
                    <center>
                        <li class="d-flex flex-dc">
                            <h3>URLs</h3>
                            <div class="tableau">
                                <table>
                                <select id="status_code">
                                  <option value="-1">All</option>
                                  <option value="plus">200</option>
                                  <option value="redirect">301/302</option>
                                  <option value="forbi">403</option>
                                </select>
                                <br>
                                    <tr>
                                      <td>Date</td>
                                      <td>Url</td>
                                      <td>Status</td>
                                        {}
                                    </tr>
                                  </table>
                            </div>
                        </li>
                        <li class="d-flex flex-dc">
                            <h3>MAIL</h3>
                            <div class="tableau">
                                <table>
                                    <tr>
                                      <td>Mails</td>
                                      <td>Status</td>
                                        {}
                                    </tr>
                                  </table>
                            </div>
                        </li>
                        <li class="d-flex flex-dc">
                            <h3>Links</h3>
                            <div class="tableau">
                                <table>
                                    <tr>
                                      {}
                                    </tr>
                                  </table>
                            </div>
                        </li>
                        <li class="d-flex flex-dc">
                            <h3>Wayback</h3>
                            <div class="tableau">
                                <table>
                                    <tr>
                                      {}
                                    </tr>
                                  </table>
                            </div>
                        </li>
                    </center>
                    </ul>
                </div>

                <!-- <div class="w1-4 centerText">
                    <h2>Test Subtitle</h2>
                </div>
                <div class="w1-4 centerText">
                    <h2>Test Subtitle</h2>
                </div>
                <div class="w1-4 centerText">
                    <h2>Test Subtitle</h2>
                </div>
                <div class="w1-4 centerText">
                    <h2>Test Subtitle</h2>
                </div> -->
            </div>
        </main>

        
    </body>
</html>'''.format(waf, cms, auth_stat, urls, mails, link, wayback))

"""if __name__ == '__main__':
    directory = "/sites/etigo/"
    cookie_ = None
    create_report(directory, cookie_)"""#DEBUG

modules/parsing_html.py
from bs4 import BeautifulSoup
import requests
import csv
import sys, re, os
from config import S3
import traceback

requests.packages.urllib3.disable_warnings(requests.packages.urllib3.exceptions.InsecureRequestWarning)

urls_s3 = []

class parsing_html:
    """
    Parsing_html: class with all function who parse html
    """
    def get_links(self, req, directory):
        """
        Get_links: get all links on webpage during the scan
        """
        soup = BeautifulSoup(req.text, "html.parser")
        search = soup.find_all('a')
        if search:
            for s in search:
                link = s.get("href")
                try:
                    if re.match(r'http(s)', link):
                        with open(directory + "/links.txt", "a+") as links:
                            links.write(str(link+"\n"))
                    else:
                        pass
                except:
                    pass


    def search_s3(self, res, req, directory):
        """
        search_s3: Check on source page if a potentialy "s3 amazon bucket" is there
        """
        s3_keyword = ["S3://", "s3-", "amazonaws"]
        for s3_f in s3_keyword:
            reqtext = req.text.split(" ")
            for req_key in reqtext:
                req_value = req_key.split('"')
                for r in req_value:
                    if s3_f in r:
                        if not os.path.exists(directory + "/s3_links.txt"):
                            with open(directory + "/s3_links.txt", "a+") as s3_links:
                                s3_links.write(str(r+"\n"))
                        else:
                            with open(directory + "/s3_links.txt", "a+") as read_links:
                                for rl in read_links.readlines():
                                    if r == rl:
                                        pass
                                    else:
                                        try:
                                            req_s3 = requests.get(r, verify=False)
                                            if req_s3.status_code == 200:
                                                print("{} Potentialy s3 buckets found with reponse 200: {}".format(S3, r))
                                                read_links.write(r)
                                        except:
                                            pass


    def mail(self, req, directory, all_mail):
        """
        Mail:
        get mail adresse in web page during the scan and check if the mail leaked
        """
        mails = req.text
        # for all @mail
        reg = re.compile(r"[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+")
        search = re.findall(reg, mails)
        for mail in search:
            #check if email pwned
            if mail and not "png" in mail or not "jpg" in mail or not "jpeg" in mail:
                datas = { "act" : mail, "accounthide" : "test", "submit" : "Submit" }
                req_ino = requests.post("https://www.inoitsu.com/", data=datas, verify=False)
                if "DETECTED" in req_ino.text:
                    pwnd = "{}: pwned ! ".format(mail)
                    if pwnd not in all_mail:
                        all_mail.append(pwnd)
                else:
                    no_pwned = "{}: no pwned ".format(mail)
                    if no_pwned not in all_mail:
                        all_mail.append(no_pwned)
        with open(directory + '/mail.csv', 'a+') as file:
            if all_mail is not None and all_mail != []:
                writer = csv.writer(file)
                for r in all_mail:
                    r = r.split(":")
                    writer.writerow(r)

    def sitemap(self, req, directory):
        """Get sitemap.xml of website"""
        soup = BeautifulSoup(req.text, "html.parser")
        with open(directory + '/sitemap.xml', 'w+') as file:
            file.write(str(soup).replace(' ','\n'))

modules/mini_scans.py
import requests
from bs4 import BeautifulSoup
import whois
import json
import sys, re, os
import ssl, OpenSSL
import socket
import pprint
import traceback
# External
from config import PLUS, WARNING, INFO, LESS, LINE, FORBI, BACK

class mini_scans:

    def get_header(self, url, directory):
        """Get header of website (cookie, link, etc...)"""
        r = requests.get(url, allow_redirects=False, verify=False)
        head = r.headers
        print(INFO + "HEADER")
        print(LINE)
        print("  {} \n".format(head).replace(',','\n'))
        print(LINE)
        with open(directory + '/header.csv', 'w+') as file:
            file.write(str(head).replace(',','\n'))


    def gitpast(self, url):
        """
        Github: check github informations
        Pastebin: check pastebin information #TODO
        """
        print("{}Check in Github".format(INFO))
        print(LINE)
        if "www" in url:
            url = url.split(".")[1]
        else:
            url = url.split("/")[2]
        url = "{}".format(url)
        print("search: {}\n".format(url))
        types = ["Commits", "Issues", "Code", "Repositories", "Marketplace", "Topics", "Wikis", "Users"]
        for t in types:
            github = "https://github.com/search?q={}&type={}".format(url, t)
            req = requests.get(github, verify=False)
            soup = BeautifulSoup(req.text, "html.parser")
            search = soup.find('a', {"class":"menu-item selected"})
            if search:
                for s in search.find("span"):
                    print("  {}{}: {}".format(INFO, t, s))
            else:
                print("  {}{}: not found".format(INFO, t))
        print("\n" + LINE)


    def who_is(self, url, directory):
        """Get whois of website"""
        print(INFO + "WHOIS")
        print(LINE)
        try:
            who_is = whois.whois(url)
            #pprint.pprint(who_is + "\n")
            for k, w in who_is.iteritems():
                is_who = " {} : {}-".format(k, w)
                print(is_who)
                with open(directory + '/whois.csv', 'a+') as file:
                    file.write(is_who.replace("-","\n"))
        except:
            erreur = sys.exc_info()
            typerr = u"%s" % (erreur[0])
            typerr = typerr[typerr.find("'")+1:typerr.rfind("'")]
            print(typerr)
            msgerr = u"%s" % (erreur[1])
            print(msgerr)
        print("\n" + LINE)


    def get_dns(self, url, directory):
        """Get DNS informations"""
        port = 0
        try:
            if "https" in url:
                url = url.replace('https://','').replace('/','')
                port = 443
            else:
                url = url.replace('http://','').replace('/','')
                port = 80
            context = ssl.create_default_context()
            conn = context.wrap_socket(socket.socket(socket.AF_INET), server_hostname=url)
            conn.connect((url, port))
            cert = conn.getpeercert()
            print(INFO + "DNS information")
            print(LINE)
            pprint.pprint(str(cert['subject']).replace(',','').replace('((','').replace('))',''))
            pprint.pprint(cert['subjectAltName'])
            print('')
            conn.close()
            print(LINE)
            with open(directory + '/dns_info.csv', 'w+') as file:
                file.write(str(cert).replace(',','\n').replace('((','').replace('))',''))
        except:
            print(INFO + "DNS information")
            print(LINE)
            erreur = sys.exc_info()
            typerr = u"%s" % (erreur[0])
            typerr = typerr[typerr.find("'")+1:typerr.rfind("'")]
            print(typerr)
            msgerr = u"%s" % (erreur[1])
            print(msgerr + "\n")
            print(LINE)


    def firebaseio(self, url):
        """
        Firebaseio: To check db firebaseio
        ex: --firebase facebook
        """
        get_domain = url.split("/")[2]
        parse_domain = get_domain.split(".")
        if not "www" in get_domain:
            if len(parse_domain) > 2:
                dire = "{}-{}".format(parse_domain[0], parse_domain[1])
            else:
                dire = "{}".format(parse_domain[0])
        else:
            dire = "{}".format(parse_domain[1])
        print("{}Firebaseio Check".format(INFO))
        print(LINE)
        url = 'https://{}.firebaseio.com/.json'.format(dire.split(".")[0])
        print(url + "\n")
        r = requests.get(url, verify=False).json()
        try:
            if 'error' in r.keys():
                if r['error'] == 'Permission denied':
                    print("  {}{} seems to be protected".format(FORBI, url)) #successfully protected
                elif r['error'] == '404 Not Found':
                    print("  {}{} not found".format(LESS, url)) #doesn't exist
                elif "Firebase error." in r['error']:
                    print("  {}{} Firebase error. Please ensure that you spelled the name of your Firebase correctly ".format(WARNING, url))
            else:
                print("  {}{} seems to be vulnerable !".format(PLUS, url)) #vulnerable
        except AttributeError:
            '''
            Some DBs may just return null
            '''
            print("{} null return".format(INFO))
        print(LINE + "\n")
        

    def wayback_check(self, url, directory):
        """
        Wayback_check:
        Check in a wayback machine to found old file on the website or other things...
        Use "waybacktool"
        """
        print("{}Wayback Check".format(INFO))
        print(LINE)
        print(url + "\n")
        try:
            os.system('python tools/waybacktool/waybacktool.py pull --host {} | python tools/waybacktool/waybacktool.py check > {}/wayback.txt'.format(url, directory))
        except Exception:
            traceback.print_exc()
        try:
            statinfo = os.path.getsize(directory + "/wayback.txt")
        except:
            print("\t{}Nothing wayback found !".format(INFO))
        if statinfo < 1 :
            print("\t{}Nothing wayback found".format(INFO))
        else:
            with open(directory + "/wayback.txt", "r+") as wayback:
                wb_read = wayback.read().splitlines()
                for wb in wb_read:
                    wb_res = list(wb.split(","))
                    try:
                        if wb_res[1] == " 200":
                            print("{}{}{}".format(PLUS, wb_res[0], wb_res[1]))
                        elif wb_res[1] == " 301" or wb_res[1] == " 302":
                            print("{}{}{}".format(LESS, wb_res[0], wb_res[1]))
                        elif wb_res[1] == " 404" or wb_res[1] == " 403":
                            pass
                        else:
                            print("{}{}{}".format(INFO, wb_res[0], wb_res[1]))
                    except:
                        pass
        print(LINE)
